### PR TITLE
ci: move the remaining travis tests to circle

### PR DESCRIPTION
**Pull request by IgorMinar**
This is the last chunk of work to move us off of Travis.

Huge thanks to @devversion for helping me work through this PR.

### CODE REVIEW

File: scripts/sauce/sauce_connect_wait_for_tunnel.sh [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
```
Comment (IgorMinar) [inline scripts/sauce/sauce_connect_wait_for_tunnel.sh]: I'm not even sure if we can use this script now that the build happens in a different job...

File: .circleci/config.yml [hunk 2]
```diff
@@ -468,14 +468,26 @@ jobs:
       - run:
           name: Building Angular benchpress output
           command: |
+            bazelBinary=$(yarn bin bazel)
```
Comment (IgorMinar) [inline .circleci/config.yml]: this variable seems to decrease readability - it's used only one in a nested expansion.

I'd prefer the following:

File: .circleci/config.yml [hunk 3]
```diff
@@ -468,14 +468,26 @@ jobs:
...
+            bazelBinOutput=$(${bazelBinary} info bazel-bin)
```
Comment (IgorMinar) [inline .circleci/config.yml]: bazelBinDir=$(yarn bazel info bazel-bin)

File: .circleci/config.yml [hunk 4]
```diff
@@ -468,14 +468,26 @@ jobs:
...
+      - run:
+          name: Starting servers for e2e tests
+          command: yarn gulp serve serve-examples
+          background: true
+
+      - run: NODE_PATH=$NODE_PATH:./dist/all $(npm bin)/protractor ./protractor-e2e.conf.js --bundles=true
+      - run: NODE_PATH=$NODE_PATH:./dist/all $(npm bin)/protractor ./protractor-examples-e2e.conf.js --bundles=true
+      - run: NODE_PATH=$NODE_PATH:./dist/all $(npm bin)/protractor ./protractor-perf.conf.js --bundles=true --dryrun
+
+      - run: yarn gulp source-map-test
```
Comment (IgorMinar) [inline .circleci/config.yml]: this one doesn't really have to do anything with e2e tests. maybe we should create a new job called `legacy-misc-tests` and move this one and cycle check in there...

File: modules/tsconfig.json [hunk 5]
```diff
@@ -12,7 +12,7 @@
     "paths": {
       "selenium-webdriver": ["../node_modules/@types/selenium-webdriver/index.d.ts"],
       "rxjs/*": ["../node_modules/rxjs/*"],
-      "@angular/*": ["../dist/all/@angular/*"]
+      "@angular/*": ["../dist/packages-dist/*"]
```
Comment (IgorMinar) [inline modules/tsconfig.json]: 👍

File: tools/gulp-tasks/check-cycle.js [hunk 6]
```diff
@@ -10,7 +10,10 @@
 module.exports = (gulp) => (done) => {
   const madge = require('madge');
 
-  const dependencyObject = madge(['dist/all/'], {
+  // TODO: This only checks for circular dependencies within each package because
+  // imports to other packages cannot be resolved by Madge when CommonJS is used.
+  // We should consider updating Madge and use a tsonfig to check across packages.
```
Comment (IgorMinar) [inline tools/gulp-tasks/check-cycle.js]: how did you discover this limitation? 

if this is true, then we should prioritize moving this test over to bazel, and in bazel creating a package-less file layout similar to `dist/all` and run the check there. (or updating madge, whatever is easier).

the reason for this tests is that in google3 we can't have cycles due to closure compiler, so it is important that we don't introduce any otherwise g3sync will fail. it's not common for us to introduce cycle especially across packages, so it's ok to degrade this test for a bit.

File: modules/benchmarks/e2e_test/tree_data.ts [hunk 7]
```diff
@@ -36,8 +36,6 @@ export const Benchmarks: Benchmark[] = [
     url: 'all/benchmarks/src/tree/ng2_next/index.html',
     buttons: CreateDestroyDetectChangesButtons,
     ignoreBrowserSynchronization: true,
-    // Can't use bundles as we use non exported code
```
Comment (IgorMinar) [inline modules/benchmarks/e2e_test/tree_data.ts]: this is no longer true. so it's ok to use bundles.

File: modules/tsconfig.json [hunk 8]
```diff
@@ -12,7 +12,7 @@
     "paths": {
       "selenium-webdriver": ["../node_modules/@types/selenium-webdriver/index.d.ts"],
       "rxjs/*": ["../node_modules/rxjs/*"],
-      "@angular/*": ["../dist/packages-dist/*"]
+      "@angular/*": ["../dist/all/@angular/*"]
```
Comment (IgorMinar) [inline modules/tsconfig.json]: why did you revert this change? I liked the way you had it. there must have been a reason but you didn't mention it in the commit message.

File: .circleci/config.yml (hunk 4 above)
Comment (devversion) [reply .circleci/config.yml]: Good point.

File: tools/gulp-tasks/check-cycle.js (hunk 6 above)
Comment (devversion) [reply tools/gulp-tasks/check-cycle.js]: I just ran Madge manually and it looked like it wasn't able to resolve imports like `@angular/core` to the proper entry point. We never specified any path mappings, so I'm fairly confident that it doesn't do that right now.

Running like `yarn madge --format cjs --extension js --summary --break-on-error dist/all` causes it to break while resolving `@angular/animations`. We should look into this, agreed, but that seems unrelated to this change, and would be better as a follow-up.

File: .circleci/config.yml [hunk 9]
```diff
@@ -462,3 +562,14 @@ workflows:
...
+
+# TODO:
+# - don't build the g3 branch
+# - verify that we are bootstrapping with the right yarn version coming from the docker image
+# - do we need node_modules check?
+#   (node tools/npm/check-node-modules --purge && yarn postinstall) || yarn install --frozen-lockfile --non-interactive
+# - add folding
+# - check local chrome version pulled from docker image
+# - remove /tools/ngcontainer
+# - could we install sauce-connect via yarn?
```
Comment (IgorMinar) [inline .circleci/config.yml]: done

File: .circleci/config.yml [hunk 10]
```diff
@@ -462,3 +562,14 @@ workflows:
             branches:
               only:
                 - master
+
+# TODO:
+# - don't build the g3 branch
+# - verify that we are bootstrapping with the right yarn version coming from the docker image
+# - do we need node_modules check?
+#   (node tools/npm/check-node-modules --purge && yarn postinstall) || yarn install --frozen-lockfile --non-interactive
+# - add folding
```
Comment (IgorMinar) [inline .circleci/config.yml]: not needed

File: .circleci/config.yml [hunk 11]
```diff
@@ -462,3 +562,14 @@ workflows:
...
+# - check local chrome version pulled from docker image
+# - remove /tools/ngcontainer
+# - could we install sauce-connect via yarn?
+# - could we have one circleci vm that would be the saucelabs tunnel for the entire workflow?
```
Comment (IgorMinar) [inline .circleci/config.yml]: not needed here, but something we might want to consider for running the unit tests under bazel...

**Comment by IgorMinar** [comment]
this PR does not affect g3: https://test.corp.google.com/ui#id=OCL:228041852:BASE:228041924:1546769193258:89b97bbd

File: .circleci/config.yml [hunk 12]
```diff
@@ -218,11 +218,12 @@ jobs:
       - run: yarn --cwd aio tools-test
       - run: ./aio/aio-builds-setup/scripts/test.sh
 
-  test_docs_examples_0:
+  test_docs_examples:
     <<: *job_defaults
     docker:
       # Needed because the example e2e tests depend on Chrome.
       - image: *browsers_docker_image
+    parallelism: 2
```
Comment (gkalpak) [inline .circleci/config.yml]: With all the changes in this PR, this is now the slowest path. Increasing parallelism here could save use another 60-90 secs :wink:)

**Review by gkalpak — DISMISSED** [review]
Woohoooo!!! 🎉 🎉 🎉 

Letf a few minor comments/suggestions, but generally LGTM 💯 
A couple more comments:
- It might be good enough for now, but the `integration_tests` sharding can be further improved to get more uniform times per container. Currently we get: 6m, 7m, 3.5m, 5.5m.
- This PR also fixes #21707 and #27663. Please, add that info in the appropriate commits.

File: modules/build.sh [hunk 13]
```diff
@@ -0,0 +1,47 @@
+#!/bin/bash
+
+# Immediately exit if any command failed.
+set -e
```
[the remark this answers is not in the document]
Comment (IgorMinar) [reply modules/build.sh]: yes. please use the same prefix as what we use everywhere:

```
set -x -u -e -o pipefail
```

File: modules/build.sh [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+
+# Go to project directory.
+cd $(dirname ${0})/..
```
[the remark this answers is not in the document]
Comment (IgorMinar) [reply modules/build.sh]: what other script does this wrapping? I'm not aware.

typically we do the following:

https://github.com/angular/angular/blob/master/scripts/ci/build.sh#L6

and we use the var to resolve paths and not rely on CWD

File: .circleci/config.yml (hunk 12 above)
Comment (IgorMinar) [reply .circleci/config.yml]: we can give it a shot. long term all these tests should run via bazel on RBE and we should not need to use circle parallelism.

one thing to keep in mind is that a big downside of circle's paralellism feature is that it's easy to incorrectly allocate resources and either under-allocate or over-allocate VMs. with bazel and rbe we are getting automatic resource allocation that doesn't need to be closely managed and monitored.

File: integration/get-sharded-tests.js [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+const manualTestShards = [
+  // The first CircleCI container only runs the bazel integration tests because these take up
+  // a lot of time and shouldn't be split up automatically.
+  ['bazel', 'bazel-schematics']
+];
+
+// Current node index of the CircleCI job. A CircleCI job can contain multiple nodes if the
+// parallelism option is enabled. This value is used to determine the tests that need to be
+// run for this instance.
+const containerInstanceIndex = Number(process.env['CIRCLE_NODE_INDEX']);
```
[the remark this answers is not in the document]
Comment (IgorMinar) [reply integration/get-sharded-tests.js]: SGTM

File: integration/get-sharded-tests.js [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+ * Splits the specified tests into a limited amount of shards. The shards of tests are being
+ * created deterministically and therefore we get reproducible tests when executing the same
+ * CircleCI instances multiple times.
+ */
+function splitTestsIntoShards(tests, maxInstances) {
+  let shards = [];
+  let assignedTests = 0;
+
+  for (let i = 0; i < maxInstances; i++) {
+    const shardSize = Math.floor((tests.length - assignedTests) / (maxInstances - i));
```
[the remark this answers is not in the document]
Comment (IgorMinar) [reply integration/get-sharded-tests.js]: all shards start in parallel as far as I can tell so `ceil` will have no significant impact.

[the remark this answers is not in the document]
Comment (IgorMinar) [reply integration/get-sharded-tests.js]: @gkalpak module approach doesn't take into account that some integration tests take much longer and need their own special shard. I think Pauls current approach is good and eventually we just need to replace all of this with bazel+rbe and removing the sharding.

File: docs/DEVELOPER.md [hunk 17]
```diff
@@ -82,7 +82,7 @@ Before submitting a PR, do not forget to remove them:
 To build Angular run:
 
 ```shell
-./build.sh
+./scripts/build-packages-dist.sh
```
[the remark this answers is not in the document]
Comment (IgorMinar) [reply docs/DEVELOPER.md]: +1

File: integration/get-sharded-tests.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (devversion) [reply integration/get-sharded-tests.js]: > all shards start in parallel as far as I can tell so ceil will have no significant impact.

Agreed on that point. Though if we make this "CI-agnostic", it would be better to use "ceil" in case there is a container queue.

>  module approach doesn't take into account that some integration tests take much longer and need their own special shard. I think Pauls current approach is good and eventually we just need to replace all of this with bazel+rbe and removing the sharding.

I initially thought the same and that's why I didn't use the `--shards {moduloResult}/{moduloDivider}` approach, but I think I got it working with modulo and it feels more clean and less verbose.

File: integration/get-sharded-tests.js [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+ * The output of this script can then be used to only run the tests which are assigned to the
+ * current CircleCI container.
+ */
+
+const fs = require('fs');
+const path = require('path');
+
+// List of all integration tests that are available.
+const integrationTests =
+    fs.readdirSync(path.join(__dirname))
```
[the remark this answers is not in the document]
Comment (devversion) [reply integration/get-sharded-tests.js]: Good catch 😄

File: scripts/saucelabs/wait-for-tunnel.sh [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+
+# Waits for Saucelabs Connect to be ready before executing any tests.
+counter=0
+
+while [[ ! -f ${SAUCE_READY_FILE} ]]; do
+  counter=$((counter + 1))
+
+  # Counter needs to be multiplied by two because the while loop only sleeps a half second.
+  # This has been made in favor of better progress logging (printing dots every half second)
+  if [ $counter -gt $[${SAUCE_READY_FILE_TIMEOUT} * 2] ]; then
```
[the remark this answers is not in the document]
Comment (devversion) [reply scripts/saucelabs/wait-for-tunnel.sh]: I don't think that changes too often and rather makes the script more complicated than it needs to be.

File: scripts/saucelabs/wait-for-tunnel.sh [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+    echo ""
+    echo "Timed out after 2 minutes waiting for tunnel ready file"
```
[the remark this answers is not in the document]
Comment (devversion) [reply scripts/saucelabs/wait-for-tunnel.sh]: Good idea

File: .circleci/config.yml [hunk 21]
```diff
@@ -444,6 +444,48 @@ jobs:
...
+        # Build the e2e tests using the existing Bazel "packages-dist" output that has been
+        # attached to this job. This avoids multiple rebuilds across various CI jobs.
+      - run: ./scripts/build-e2e-tests.sh --use-existing-packages-dist
+      - run:
+          name: Starting servers for e2e tests
+          command: yarn gulp serve serve-examples
+          background: true
+      - run: NODE_PATH=$NODE_PATH:./dist/all $(npm bin)/protractor ./protractor-e2e.conf.js --bundles=true
+      - run: NODE_PATH=$NODE_PATH:./dist/all $(npm bin)/protractor ./protractor-examples-e2e.conf.js --bundles=true
+      - run: NODE_PATH=$NODE_PATH:./dist/all $(npm bin)/protractor ./protractor-perf.conf.js --bundles=true --dryrun
```
[the remark this answers is not in the document]
Comment (devversion) [reply .circleci/config.yml]: It technically does. Running with yarn usually brings a modified shell environment. Not sure if this is a problem here, so I'll try running with `yarn`.

File: .circleci/config.yml [hunk 22]
```diff
@@ -444,6 +444,48 @@ jobs:
...
+          key: *cache_key
+      - *define_env_vars
+      - *yarn_install
+      - attach_workspace:
+          at: dist
+
+      - run: yarn gulp check-cycle
+      # TODO: disabled because the Bazel packages-dist does not seem to have map files for
+      # the ESM5/ES2015 output.
+      # - run: yarn gulp source-map-test
```
[the remark this answers is not in the document]
Comment (devversion) [reply .circleci/config.yml]: There wasn't. I've just created a Github issue for tracking. https://github.com/angular/angular-cli/issues/13378

File: .circleci/env.sh [hunk 23]
```diff
@@ -31,5 +34,26 @@ setSecretVar CI_SECRET_PAYLOAD_FIREBASE_TOKEN "$ANGULAR_PAYLOAD_TOKEN";
 setSecretVar CI_SECRET_SLACK_CARETAKER_WEBHOOK_URL "$SLACK_CARETAKER_WEBHOOK_URL";
 
 
+####################################################################################################
+# Define SauceLabs environment variables for CircleCI.
+####################################################################################################
+# In order to have a meaningful SauceLabs badge on the repo page,
+# the angular2-ci account is used only when pushing commits to master;
+# in all other cases, the regular angular-ci account is used.
+if [ "${CI_PULL_REQUEST}" = "false" ] && [ "${CI_BRANCH}" = "master" ]; then
```
[the remark this answers is not in the document]
Comment (devversion) [reply .circleci/env.sh]: This should be guarded by the `CI_PULL_REQUEST` check. Additionally if Igor makes the access key a secret CircleCI variable, we should be guarded as well.

File: .circleci/env.sh [hunk 24]
```diff
@@ -31,5 +34,26 @@ setSecretVar CI_SECRET_PAYLOAD_FIREBASE_TOKEN "$ANGULAR_PAYLOAD_TOKEN";
...
+  setPublicVar SAUCE_USERNAME "angular2-ci";
+  setSecretVar SAUCE_ACCESS_KEY "693ebc16208a-0b5b-1614-8d66-a2662f4e";
```
[the remark this answers is not in the document]
Comment (devversion) [reply .circleci/env.sh]: cc. @IgorMinar. I think this is reasonable. This would *ensure* that the Saucelabs badge is 100% based on the results of the `angular/angular#master` branch.

File: docs/DEVELOPER.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (devversion) [reply docs/DEVELOPER.md]: That's true, but what would be the proper message if we only remove `build.sh`? (if we only consider 2431a97e40793ca65dee6094fc8777819d13b662)

I will just swap the commits, so that we remove `test.sh` before `build.sh`. It should be fine then.

File: docs/DEVELOPER.md [hunk 25]
```diff
@@ -85,30 +85,20 @@ To build Angular run:
...
-See [DEBUG.md](DEBUG.md) for information on debugging the code while running the unit tests.
-
-All the tests are executed on our Continuous Integration infrastructure and a PR could only be merged once the tests pass.
+Read more about this in the [BAZEL.md](./BAZEL.md) document. You should execute all test suites
+before submitting a PR to Github.
 
-- CircleCI fails if your code is not formatted properly,
-- Travis CI fails if any of the test suites described above fails.
+All the tests are executed on our Continuous Integration infrastructure and a PR could only be
+merged once all tests pass. CircleCI fails if your code is not is passing all tests or is not
```
[the remark this answers is not in the document]
Comment (devversion) [reply docs/DEVELOPER.md]: Agreed. I basically just based it on what was there before 😄 CircleCI and Travis were mentioned at this point.

File: .circleci/config.yml (hunk 21 above)
[the remark this answers is not in the document]
Comment (gkalpak) [reply .circleci/config.yml]: If there is a specific reason not to use `yarn protractor`, I would consider leaving a comment explaining that.
I would also use `yarn bin` over `npm bin`, because we are generally using yarn :wink:

File: .circleci/config.yml (hunk 12 above)
Comment (devversion) [reply .circleci/config.yml]: Agreed. I've just switched it to three parallel instances for now.

File: scripts/saucelabs/start-tunnel.sh [hunk 26]
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
```
[the remark this answers is not in the document]
Comment (devversion) [reply scripts/saucelabs/start-tunnel.sh]: Sure. Makes sense. Also there are a lot of other scripts in the while repo that use `!/bin/bash`. We should try to switch them over at some point (out of scope for this PR)

File: modules/build.sh (hunk 14 above)
[the remark this answers is not in the document]
Comment (gkalpak) [reply modules/build.sh]: > what other script does this wrapping? I'm not aware.

@IgorMinar, you introduced this "concept" in b8f0c3dc7 ([example 1][1], [example 2][2], [example 3][3]).
But maybe my understanding of the reasoning is incorrect.

Just to be clear, I was referring to changing this:

```
cd some/dir/
run some code
```

to that:

```
(
  cd some/dir/
  run some code
)
```

##
@devversion, good point about `$0`. It will indeed not work as expected when sourced.
I am either misunderstanding the point of wrapping in `(...)` or there is no point here anyway.

In any case, I don't feel strongly at all about.

[1]: https://github.com/angular/angular/commit/b8f0c3dc7bdf37553fa564578635f5bfcc6c646c#diff-cf6136203c5f24fba45e4ab92e7895e3R52
[2]: https://github.com/angular/angular/commit/b8f0c3dc7bdf37553fa564578635f5bfcc6c646c#diff-241beae81a9a4b50c44ecde82e496324R11
[3]: https://github.com/angular/angular/commit/b8f0c3dc7bdf37553fa564578635f5bfcc6c646c#diff-5ead0d120428a4b905965643dac1af0eR11

File: .circleci/env.sh (hunk 23 above)
[the remark this answers is not in the document]
Comment (devversion) [reply .circleci/env.sh]: True, it would guard against that if they don't update their setup. Best would be if Igor makes this is a secret CircleCI key. I'll add the safety check according to your feedback anyway 👍

File: modules/build.sh (hunk 14 above)
[the remark this answers is not in the document]
Comment (devversion) [reply modules/build.sh]: Using the `(...)` basically makes this execute in a subshell, so that `cd` does not have an effect on the primary shell. 

Though since we don't support running this script with `source`, the directory change will never affect the main shell because the "normal running" of the script also runs in a subshell.

File: .circleci/config.yml [hunk 27]
```diff
@@ -462,3 +513,9 @@ workflows:
             branches:
               only:
                 - master
+
+# TODO:
+# - don't build the g3 branch
+# - verify that we are bootstrapping with the right yarn version coming from the docker image
+# - check local chrome version pulled from docker image
+# - remove /tools/ngcontainer
```
[the remark this answers is not in the document]
Comment (devversion) [reply .circleci/config.yml]: @gkalpak It's not planned to address this as part of this PR, unless @IgorMinar wants me to do that.

I think tracking this fits better into Jira, so you someone who has access needs to create one. Maybe Alex already has something on this list?

File: modules/build.sh (hunk 14 above)
[the remark this answers is not in the document]
Comment (IgorMinar) [reply modules/build.sh]: @gkalpak right. I did that, but it only makes sense AFAIK if you cd as part of a portion of a script. doing it at the beginning of a script doesn't really make a difference because CWD will be reset after the script exists.

[the remark this answers is not in the document]
Comment (gkalpak) [reply modules/build.sh]: Right. So, it only makes sense if you want to source the script or have multiple `cd`s in the same file (neither of which is true here).
Thanks for clarifying 👍

**Review by alexeagle — APPROVED** [review]
I think developer.md needs a bunch more work, but maybe in a follow-up PR since this is close to landing. 
should just be `yarn build`, `yarn test`, `yarn e2e` and I don't see why ppl want to wait to build the real npm packages, what's the workflow for that?
Let's go through it together?

File: .circleci/config.yml (hunk 27 above)
[the remark this answers is not in the document]
Comment (gkalpak) [reply .circleci/config.yml]: @devversion, sorry for the late response. Are you (or anyone else) working on these or should we create a Jira issue to track?

[the remark this answers is not in the document]
Comment (devversion) [reply .circleci/config.yml]: @gkalpak No problem 😄 I'm currently not working on these, nor we have these tracked anywhere yet (except in this file).

Agreed, I think it would be good to move these to JIRA. I personally would like to be involved with the removal of `ngcontainer` since we still use it on Material and I've been trying to remove it already in #26691. If you create one, can you mention that Material still depends on it atm?

**Comment by IgorMinar** [comment]
We can't remove ngcontainer yet. I talked to AlexE. Too many other repos
need it.
